### PR TITLE
RealmCollectionType

### PR DESCRIPTION
**Pull request by jpsim**
/cc @segiddins

### CODE REVIEW

**Comment by segiddins** [comment]
What about `realm`, `invalidated, and KVC?

**Comment by jpsim** [comment]
* `realm` is optional on `List`
* `invalidated` can't happen to a `Results`
* `KVC` is defined on `NSObject`

**Comment by segiddins** [comment]
- it should be fine to define it as an optional on the protocol but make it non-optional on `Results`
- it actually can, if the realm is invalidated
- true, but `RealmCollectionType` isn't limited to `NSObject` types, as far as I can see

**Comment by jpsim** [comment]
> it should be fine to define it as an optional on the protocol but make it non-optional on `Results`

That doesn't work.

> it actually can, if the realm is invalidated

We don't expose that at the moment. Create a new issue if you think we should in both Objective-C and Swift results.

> true, but `RealmCollectionType` isn't limited to `NSObject` types, as far as I can see

Ugh, I can't make `RealmCollectionType` conform to `NSObject` or `@objc` because then it couldn't also conform to `CollectionType`. I could make a `KeyValueEncodable` protocol I suppose...

**Comment by astigsen** [comment]
>> What about `realm`?
> `realm` is optional on `List`

We definitely need some way to access the realm from a `RealmCollectionType`. Otherwise you can't do any transactions on it without having to communicate the realm off-band.

Would it be enough to make the `realm` variable optional in `Results` as well?

**Comment by jpsim** [comment]
> Would it be enough to make the realm variable optional in Results as well?

We could, but that would slightly weaken the interface on `Results`. I suppose that's a worthwhile tradeoff, though. I'll do that.

**Comment by jpsim** [comment]
Done.

File: RealmSwift-swift1.2/RealmCollectionType.swift [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+////////////////////////////////////////////////////////////////////////////
+
+import Foundation
+import Realm
+
+/**
+Encapsulates iteration state and interface for iteration over a
+`RealmCollectionType`.
+*/
+public class RLMGenerator<T: Object>: GeneratorType {
```
Comment (segiddins) [inline RealmSwift-swift1.2/RealmCollectionType.swift]: `final`?

File: RealmSwift-swift1.2/RealmCollectionType.swift [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    */
+    func indexOf(predicateFormat: String, _ args: CVarArgType...) -> Int?
+
+
+    // MARK: Object Retrieval
+
+    /**
+    Returns the object at the given `index`.
+
+    :warning: You can only set an object during a write transaction.
```
Comment (segiddins) [inline RealmSwift-swift1.2/RealmCollectionType.swift]: not true for standalone lists

File: RealmSwift-swift1.2/RealmCollectionType.swift [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+}
+
+/**
+A homogenous collection of `Object`s which can be retrieved, filtered, sorted,
+and operated upon.
+*/
+public protocol RealmCollectionType: CollectionType {
+
+    /// Element type contained in this collection.
+    typealias Element: Object
```
Comment (segiddins) [inline RealmSwift-swift1.2/RealmCollectionType.swift]: should the type alias be public?

File: RealmSwift-swift2.0/RealmCollectionType.swift [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    // MARK: Index Retrieval
+
+    /**
+    Returns the index of the given object, or `nil` if the object is not in the collection.
+
+    - parameter object: The object whose index is being queried.
+
+    - returns: The index of the given object, or `nil` if the object is not in the collection.
+    */
+    func indexOf(object: Element) -> Int?
```
Comment (segiddins) [inline RealmSwift-swift2.0/RealmCollectionType.swift]: already implemented for equatable in swift 2, so this is redundant

File: RealmSwift-swift2.0/RealmCollectionType.swift [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    /**
+    Returns the object at the given `index`.
+
+    - warning: You can only set an object during a write transaction.
+
+    - parameter index: The index.
+
+    - returns: The object at the given `index`.
+    */
+    subscript(index: Int) -> Element { get }
```
Comment (segiddins) [inline RealmSwift-swift2.0/RealmCollectionType.swift]: already on `CollectionType`

File: RealmSwift-swift2.0/RealmCollectionType.swift [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+    /// Returns the first object in the collection, or `nil` if empty.
+    var first: Element? { get }
```
Comment (segiddins) [inline RealmSwift-swift2.0/RealmCollectionType.swift]: already on CollectionType

File: RealmSwift-swift2.0/RealmCollectionType.swift [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+    /// Returns the last object in the collection, or `nil` if empty.
+    var last: Element? { get }
```
Comment (segiddins) [inline RealmSwift-swift2.0/RealmCollectionType.swift]: already on CollectionType

File: RealmSwift-swift1.2/RealmCollectionType.swift (hunk 1 above)
Comment (jpsim) [inline RealmSwift-swift1.2/RealmCollectionType.swift]: Sure.

File: RealmSwift-swift1.2/RealmCollectionType.swift (hunk 2 above)
Comment (jpsim) [inline RealmSwift-swift1.2/RealmCollectionType.swift]: I missed this.

File: RealmSwift-swift1.2/RealmCollectionType.swift (hunk 3 above)
Comment (jpsim) [inline RealmSwift-swift1.2/RealmCollectionType.swift]: I don't see how that's a problem.

File: RealmSwift-swift2.0/RealmCollectionType.swift (hunk 4 above)
Comment (jpsim) [inline RealmSwift-swift2.0/RealmCollectionType.swift]: It's not redundant as this is an optimized implementation.

Comment (jpsim) [inline RealmSwift-swift2.0/RealmCollectionType.swift]: Sorry, you're right that it's redundant in the interface.

**Comment by jpsim** [comment]
@segiddins addressed your feedback. I think the only things left to do are to test aggregate operations on `List`. Marking WIP until that's done.

**Comment by segiddins** [comment]
@jpsim anything we can do to help get this in for the next release?

**Comment by jpsim** [comment]
@tgoyne @segiddins this should now be ready for review.

**Comment by segiddins** [comment]
I think some of the `List` tests are now redundant?

**Comment by jpsim** [comment]
> I think some of the List tests are now redundant?

Good point, fixed in last commit.

File: RealmSwift-swift1.2/Results.swift [hunk 8]
```diff
@@ -78,10 +78,16 @@ Results cannot be created directly.
...
 public final class Results<T: Object>: ResultsBase {
 
+    /// Element type contained in this collection.
+    typealias Element = T
+
     // MARK: Properties
 
     /// Returns the Realm these results are associated with.
-    public var realm: Realm { return Realm(rlmResults.realm) }
+    /// Despite returning an `Optional<Realm>` in order to conform to
```
Comment (tgoyne) [inline RealmSwift-swift1.2/Results.swift]: Can it conform to the protocol by returning an implicitly-unwrapped optional?

Comment (jpsim) [inline RealmSwift-swift1.2/Results.swift]: Unfortunately not.

Comment (segiddins) [inline RealmSwift-swift1.2/Results.swift]: Filed as rdar://22426323

**Comment by segiddins** [comment]
:+1:

File: RealmSwift-swift1.2/RealmCollectionType.swift [hunk 9]
```diff
@@ -0,0 +1,655 @@
...
+    func average<U: AddableType>(property: String) -> U? { fatalError() }
+    subscript(index: Int) -> Element { fatalError() }
+    func generate() -> RLMGenerator<T> { fatalError() }
+    var startIndex: Int { fatalError() }
+    var endIndex: Int { fatalError() }
+    func valueForKey(key: String) -> AnyObject? { fatalError() }
+    func setValue(value: AnyObject?, forKey key: String) { fatalError() }
+}
+
+private class _AnyRealmCollection<C: RealmCollectionType>: _AnyRealmCollectionBase<C.Element> {
```
Comment (tgoyne) [inline RealmSwift-swift1.2/RealmCollectionType.swift]: final

**Comment by tgoyne** [comment]
:+1:

**Pull request merged**